agent/api_manager/osv_service_api.py

```diff
@@ -0,0 +1,194 @@
+import dataclasses
+import json
+import logging
+from typing import Optional, Iterator
+
+import requests
+from ostorlab.agent.kb import kb
+from ostorlab.agent.mixins import agent_report_vulnerability_mixin
+
+from agent import cve_service_api
+from agent import osv_output_handler
+
+logger = logging.getLogger(__name__)
+
+OSV_ENDPOINT = "https://api.osv.dev/v1/query"
+
+RISK_RATING_MAPPING = {
+    "POTENTIALLY": agent_report_vulnerability_mixin.RiskRating.POTENTIALLY,
+    "LOW": agent_report_vulnerability_mixin.RiskRating.LOW,
+    "MEDIUM": agent_report_vulnerability_mixin.RiskRating.MEDIUM,
+    "MODERATE": agent_report_vulnerability_mixin.RiskRating.MEDIUM,
+    "HIGH": agent_report_vulnerability_mixin.RiskRating.HIGH,
+    "CRITICAL": agent_report_vulnerability_mixin.RiskRating.CRITICAL,
+}
+
+
+@dataclasses.dataclass
+class VulnData:
+    risk: str
+    description: str
+    summary: str
+    fixed_version: str | None
+    cvss_v3_vector: str | None
+    references: list[dict[str, str]]
+    cves: list[str]
+
+
+def query_osv_api(
+    package_name: str | None, version: str | None, ecosystem: str | None
+) -> Optional[str]:
+    """Query the OSv API with the specified version, package name, and ecosystem.
+    Args:
+        version: The version to query.
+        package_name: The name of the package to query.
+        ecosystem: The ecosystem of the package e.g., javascript.
+    Returns:
+        The API response text if successful, None otherwise.
+    """
+    if version is None:
+        logger.error("Error: Version must not be None.")
+        return None
+    if package_name is None:
+        logger.error("Error: Package name must not be None.")
+        return None
+
+    data = {
+        "version": version,
+        "package": {"name": package_name, "ecosystem": ecosystem},
+    }
+    headers = {"Content-Type": "application/json"}
+    response = requests.post(OSV_ENDPOINT, data=json.dumps(data), headers=headers)
+
+    if response.status_code == 200:
+        return response.text
+    else:
+        logger.error(f"Error: Request failed with status code {response.status_code}")
+        return None
+
+
+def parse_output(
+    api_response: Optional[str], api_key: str | None = None
+) -> list[VulnData]:
+    """Parse the OSv API response to extract vulnerabilities.
+    Args:
+        api_response: The API response text.
+    Returns:
+        Parsed output.
+    """
+    if api_response is None:
+        logger.error("Error: API response must not be None.")
+        return []
+
+    try:
+        response_data = json.loads(api_response)
+        vulnerabilities = response_data.get("vulns", [])
```

Comment:
Keep except json.JSONDecodeError only around the part that may raise it!

Reply:
It is. There is `json.loads(api_response)` and also `get_cve_data_from_api` that might raise it.